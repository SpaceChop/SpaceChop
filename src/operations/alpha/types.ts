import * as t from 'runtypes';
import Opacity from '../../types/Opacity';
import Param from '../../types/Param';

const AlphaConfig = t.Record({
  opacity: Opacity.Or(Param),
});

export type AlphaConfig = t.Static<typeof AlphaConfig>;
export default AlphaConfig;
