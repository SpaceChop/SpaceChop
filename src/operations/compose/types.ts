import * as t from 'runtypes';
import Arrangement from '../../types/Arrangement';
import Blending from '../../types/Blending';
import Composition from '../../types/Composition';
import Gravity from '../../types/Gravity';
import Offset from '../../types/Offset';
import Param from '../../types/Param';

const ComposeConfig = t.Record({
  source: t.String,
}).And(t.Partial({
  params: t.Dictionary(t.String, 'string'),
  preset: t.Union(t.String, Param),
  compose: t.Union(Composition, Param),
  arrange: t.Union(Arrangement, Param),
  gravity: t.Union(Gravity, Param),
  offset: Offset,
  blend: t.Union(Blending, Param),
}));

export type ComposeConfig = t.Static<typeof ComposeConfig>;
export default ComposeConfig;
