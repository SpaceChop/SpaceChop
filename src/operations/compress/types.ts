import * as t from 'runtypes';
import Param from '../../types/Param';

const Quality = t.Number.withConstraint(
  (n) => {
    if (n < 0 || n > 100) {
      return 'quality must be between 0 and 100';
    }
    if (n % 1 !== 0) {
      return 'quality must be non decimal number';
    }
    return true;
  },
);

const CompressConfig = t.Partial({
  quality: Quality.Or(Param),
  lossy: t.Boolean,
});

export type CompressConfig = t.Static<typeof CompressConfig>;
export default CompressConfig;
