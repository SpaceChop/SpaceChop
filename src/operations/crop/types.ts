import * as t from 'runtypes';
import Gravity from '../../types/Gravity';
import Offset from '../../types/Offset';
import Param from '../../types/Param';
import Percentage from '../../types/Percentage';
import PostiveNumber from '../../types/PositiveNumber';

const CropConfig = t.Partial({
  width: t.Union(PostiveNumber, Param, Percentage),
  height: t.Union(PostiveNumber, Param, Percentage),
  gravity: t.Union(Gravity, Param),
  offset: Offset,
}).withConstraint((n) => {
  if (!(n.width || n.height)) {
    return 'missing width or height';
  }
  return true;
});

export type CropConfig = t.Static<typeof CropConfig>;
export default CropConfig;
