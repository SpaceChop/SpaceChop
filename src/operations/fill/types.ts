import * as t from 'runtypes';
import Gravity from '../../types/Gravity';
import Param from '../../types/Param';
import PostiveNumber from '../../types/PositiveNumber';

const FillConfig = t.Record({
  width: t.Union(PostiveNumber, Param),
  height: t.Union(PostiveNumber, Param),
}).And(t.Partial({
  gravity: t.Union(Gravity, Param),
}));

export type FillConfig = t.Static<typeof FillConfig>;
export default FillConfig;
