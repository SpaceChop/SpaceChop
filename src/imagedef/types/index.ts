import { Stream } from 'stream';
import ImageDefinition from '../../imagedef';
import AsyncBuffer from '../../lib/asyncBuffer';
import isGIF from './gif';
import isJPEG from './jpeg';
import isPNG from './png';
import isWEBP from './webp';

const tests = [
  isGIF,
  isJPEG,
  isPNG,
  isWEBP,
];

export default (stream: Stream): Promise<ImageDefinition> => {
  const asyncBuffer = new AsyncBuffer(stream);
  return new Promise((resolve, reject) => {
    tests.map((test) => {
      test(asyncBuffer)
        .then((definition) => {
          if (definition) {
            resolve(definition);
            asyncBuffer.destroy();
          }
        })
        .catch(reject);
    });
  });
};
