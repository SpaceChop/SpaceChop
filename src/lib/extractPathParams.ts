import pathToRegexp from 'path-to-regexp';

const cache = {};
const cacheLimit = 10000;
let cacheCount = 0;

export default (pattern) => {
  if (!pattern) {
    throw new Error('Cant compile path. No pattern provided.');
  }

  if (cache[pattern]) { return cache[pattern]; }

  const keys = [];
  pathToRegexp(pattern, keys);

  if (cacheCount < cacheLimit) {
    cache[pattern] = keys;
    cacheCount++;
  }

  return keys;
};
