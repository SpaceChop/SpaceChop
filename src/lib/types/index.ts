import { Stream } from 'stream';
import ImageDefinition from '../../imagedef';
import isGIF from './gif';
import isJPEG from './jpeg';
import isPNG from './png';
import isWEBP from './webp';

const tests = [
  isGIF,
  isJPEG,
  isPNG,
  isWEBP,
];

export default (stream: Stream): Promise<any> => new Promise((resolve) => {
  const chunks = [];
  stream.on('data', (chunk) => {
    chunks.push(chunk);
  });
  stream.on('end', async () => {
    const buffer = Buffer.concat(chunks);
    for (const test of tests) {
      const result: ImageDefinition = test(buffer);
      if (result) {
        resolve(result);
        break;
      }
    }
  });
});
