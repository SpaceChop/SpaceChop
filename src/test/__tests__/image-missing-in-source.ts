import path from 'path';
import extractPathParams from '../../lib/extractPathParams';
import { Config } from '../../types/Config';
import { requestHandler } from './../../spacechop';
import assetsFolder from './../assets/dirname';
import { Request, Response } from './../utils/expressMocks';

/**
 * Tests in this file use the full spacechop implementation by mocking Express
 * Response and Requests.
 */
describe('Image missing in source', () => {
  const p = '/:preset/:image';
  const config: Config = {
    sources: [{
      volume: {
        root: path.join(assetsFolder, ':image'),
      },
    }],
    paths: [p],
    presets: {
      t_original: {
        steps: [],
      },
    },
  };

  const sources = [
    {
      exists: jest.fn(() => Promise.resolve(false)),
      stream: jest.fn(),
    },
  ];

  const params = extractPathParams(p);
  const handler = requestHandler(config, params, sources);

  let request;
  let response;
  beforeAll(async () => {
    request = new Request();
    response = new Response();
    request.setParams(0, 't_original');
    request.setParams(1, 'grid.png');
    await handler(request, response);
  });
  it('should set status 404', () => {
    expect(response.status).toBeCalledWith(404);
    expect(response.end.mock.calls[0]).toMatchSnapshot();
  });
});
