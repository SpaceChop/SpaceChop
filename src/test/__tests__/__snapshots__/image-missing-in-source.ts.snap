// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image missing in source should set status 404 1`] = `
Array [
  "Could not find image",
]
`;
